File: backend/src/routes/stereo.js
'use strict';

const express = require('express');
const router = express.Router();
const { v4: uuidv4 } = require('uuid');
const path = require('path');
const fs = require('fs');
const authMiddleware = require('../middleware/auth');

const LIBRARY_FILE = path.join(__dirname, '..', 'data', 'library.json');
const EXERCISE_TTL = 30 * 60 * 1000;
const exerciseStore = new Map();

const STEREO_LEVELS = {
  1: { options: ['mono', 'stereo'], widths: { mono: 0, stereo: 1.0 } },
  2: { options: ['schmal', 'mittel', 'breit'], widths: { schmal: 0.2, mittel: 0.5, breit: 1.0 } },
  3: { options: ['mono', 'sehr schmal', 'schmal', 'breit', 'sehr breit'], widths: { mono: 0, 'sehr schmal': 0.15, schmal: 0.35, breit: 0.7, 'sehr breit': 1.0 } },
};

const EXPLANATIONS = {
  mono: 'Mono: Kein Stereoanteil — beide Kanäle identisch.',
  stereo: 'Stereo: Volle Stereobreite — L und R unterschiedlich.',
  schmal: 'Schmal: Wenig Stereoanteil, klingt fast mono.',
  mittel: 'Mittel: Ausgewogene Stereobreite.',
  breit: 'Breit: Deutliche Stereobreite, füllt den Raum.',
  'sehr schmal': 'Sehr schmal: Kaum Stereoanteil, fast mono.',
  'sehr breit': 'Sehr breit: Maximale Stereobreite, fast übertrieben.',
};

function loadLibrary() {
  try { return JSON.parse(fs.readFileSync(LIBRARY_FILE, 'utf8')); }
  catch { return []; }
}

function purgeExpired() {
  const now = Date.now();
  for (const [id, ex] of exerciseStore) {
    if (now > ex.expiresAt) exerciseStore.delete(id);
  }
}

// GET /api/stereo/random?level=1|2|3
router.get('/random', authMiddleware, (req, res) => {
  purgeExpired();
  const level = Math.min(3, Math.max(1, parseInt(req.query.level, 10) || 1));
  const library = loadLibrary();
  const active = library.filter(e => e.active);
  const pool = active.length > 0 ? active : library;
  if (pool.length === 0) return res.status(404).json({ error: 'Keine Audiodateien in der Library' });

  const audio = pool[Math.floor(Math.random() * pool.length)];
  const cfg = STEREO_LEVELS[level];
  const options = cfg.options;
  const correctAnswer = options[Math.floor(Math.random() * options.length)];
  const width = cfg.widths[correctAnswer];

  const exerciseId = uuidv4();
  exerciseStore.set(exerciseId, {
    correctAnswer, width, level, audioId: audio.id,
    expiresAt: Date.now() + EXERCISE_TTL,
  });

  res.json({ exerciseId, audioId: audio.id, audioUrl: `/api/library/${audio.id}/audio`, level, options, width });
});

// POST /api/stereo/evaluate
router.post('/evaluate', authMiddleware, (req, res) => {
  const { exerciseId, answer, secondsTaken } = req.body;
  const exercise = exerciseStore.get(exerciseId);
  if (!exercise) return res.status(404).json({ error: 'Übung nicht gefunden oder abgelaufen' });
  exerciseStore.delete(exerciseId);

  const correct = answer === exercise.correctAnswer;
  const timeFactor = Math.max(0.3, 1 - (secondsTaken || 0) / 45);
  const points = Math.round((correct ? 1000 : 0) * timeFactor);

  res.json({
    correct,
    correctAnswer: exercise.correctAnswer,
    width: exercise.width,
    points,
    explanation: EXPLANATIONS[exercise.correctAnswer] || '',
  });
});

module.exports = router;

File: core/src/exercise/stereo.rs
//! Stereo Width Trainer — port of backend/src/routes/stereo.js.

use crate::buffer::AudioBuffer;
use crate::dsp::stereo_width::apply_stereo_width;
use crate::exercise::common::time_factor_45;
use rand::Rng;
use serde::Serialize;

pub fn level_options(level: u8) -> &'static [(&'static str, f32)] {
    match level.clamp(1, 3) {
        1 => &[("mono", 0.0), ("stereo", 1.0)],
        2 => &[("schmal", 0.2), ("mittel", 0.5), ("breit", 1.0)],
        _ => &[
            ("mono", 0.0),
            ("sehr schmal", 0.15),
            ("schmal", 0.35),
            ("breit", 0.7),
            ("sehr breit", 1.0),
        ],
    }
}

pub fn explanation(answer: &str) -> &'static str {
    match answer {
        "mono" => "Mono: Kein Stereoanteil — beide Kanäle identisch.",
        "stereo" => "Stereo: Volle Stereobreite — L und R unterschiedlich.",
        "schmal" => "Schmal: Wenig Stereoanteil, klingt fast mono.",
        "mittel" => "Mittel: Ausgewogene Stereobreite.",
        "breit" => "Breit: Deutliche Stereobreite, füllt den Raum.",
        "sehr schmal" => "Sehr schmal: Kaum Stereoanteil, fast mono.",
        "sehr breit" => "Sehr breit: Maximale Stereobreite, fast übertrieben.",
        _ => "",
    }
}

#[derive(Clone)]
pub struct StereoExercise {
    pub correct_answer: &'static str,
    pub width_factor: f32,
    pub level: u8,
}

pub fn generate(level: u8, rng: &mut impl Rng) -> StereoExercise {
    let opts = level_options(level);
    let (answer, width) = opts[rng.gen_range(0..opts.len())];
    StereoExercise { correct_answer: answer, width_factor: width, level: level.clamp(1, 3) }
}

pub fn render(dry: &AudioBuffer, exercise: &StereoExercise) -> AudioBuffer {
    apply_stereo_width(dry, exercise.width_factor)
}

#[derive(Serialize)]
pub struct StereoResult {
    pub correct: bool,
    pub points: u32,
}

pub fn evaluate(exercise: &StereoExercise, answer: &str, seconds_taken: f32) -> StereoResult {
    let correct = answer == exercise.correct_answer;
    let points = ((if correct { 1000.0 } else { 0.0 }) * time_factor_45(seconds_taken)).round() as u32;
    StereoResult { correct, points }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;

    #[test]
    fn level1_only_yields_mono_or_stereo() {
        let mut rng = rand_chacha::ChaCha8Rng::seed_from_u64(1);
        for _ in 0..50 {
            let ex = generate(1, &mut rng);
            assert!(ex.correct_answer == "mono" || ex.correct_answer == "stereo");
        }
    }
}
